### PR TITLE
Created Storage sample app with blob listing and MSAL

### PR DESCRIPTION
**Pull request by vcolin7**
In collaboration with @anuchandy.

This depends on PR #174.

### CODE REVIEW

**Comment by vcolin7** [comment]
Marked as "Do Not Merge" as it depends on PR #174.

**Comment by anuchandy** [comment]
@bsiegel - This is a sample app to list blobs in RecylerView control using PagedListAdapter. For now the paging abstractions that could be useful across client libs are in this sample (under sample\core dir). Once we agree on paging types to be in azure-core, we can move them. 

Sample is based on #174, so first that needs to be merged before we take this.

File: samples/sample-app-storage/build.gradle [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+    buildTypes {
+        release {
+            minifyEnabled false
+            proguardFiles getDefaultProguardFile('proguard-android-optimize.txt'), 'proguard-rules.pro'
+        }
+    }
+}
+
+dependencies {
```
Comment (bsiegel) [inline samples/sample-app-storage/build.gradle]: Let's get this section sorted (editor should be able to just do a "sort lines")

File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+                    public void onSuccess(IAuthenticationResult authResult) {
+                        Log.d(TAG, "[acquireTokenSilentAsync]: Successfully authenticated.");
+                        callback.onToken(new AccessToken(authResult.getAccessToken(),
+                                toOffsetDateTime(authResult.getExpiresOn())));
+                    }
+
+                    @Override
+                    public void onError(MsalException exception) {
+                        Log.d(TAG, "[acquireTokenSilentAsync]: Authentication failed: " + exception.toString());
+                        if (exception instanceof MsalClientException) {
```
Comment (bsiegel) [inline samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: Can we add a comment here about what this type of exception means?

File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+import com.microsoft.identity.client.exception.MsalServiceException;
+import com.microsoft.identity.client.exception.MsalUiRequiredException;
+
+import org.threeten.bp.DateTimeUtils;
+import org.threeten.bp.OffsetDateTime;
+import org.threeten.bp.ZoneOffset;
+
+import java.util.Date;
+
+class MsalClient {
```
Comment (bsiegel) [inline samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: Nit: How do you feel about capitalizing this (and other MSAL-related classes) like `MSALClient`?

**Comment by bsiegel** [comment]
Seeing a lint issue that is most likely a bug:

```
android/samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/ContainerBlobsPagedListAdapter.java:94: Error: Suspicious equality check: equals() is not implemented in com.azure.android.storage.blob.models.BlobItem [DiffUtilEquals]
              return oldItem.equals(newItem);
                             ~~~~~~~~~~~~~~~
```

File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java (hunk 3 above)
Comment (anuchandy) [reply samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: Java guideline is to use pascal-case for acronyms so `Msal`.

Comment (anuchandy) [reply samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: https://azure.github.io/azure-sdk/java_design.html#java-naming-uppercase-acronyms

File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java (hunk 2 above)
Comment (anuchandy) [reply samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: it's for general local error in msal library. 

@vcolin7 could you please add a comment referring to https://javadoc.io/doc/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalClientException.html

**Comment by anuchandy** [comment]
@vcolin7 for `oldItem.equals(newItem)` linting, we could just compare the {oldItem.name, newItem.name} and {oldItem.versionId, newItem.versionId} given blob names are unique within a container.

File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java (hunk 2 above)
Comment (vcolin7) [reply samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: Sure thing

**Comment by vcolin7** [comment]
@bsiegel @anuchandy Added a link to the MsalException JavaDoc, fixed the lint error with Anu's suggestion and reordered the dependencies in our Gradle file. Also solved a small merge conflict.

**Comment by anuchandy** [comment]
@vcolin7 thank you!, 2 minor feedbacks:

1. can we move the Exception doc reference to inside the `if`

```java
if (exception instanceof MsalClientException) {
    // HERE
}
```

2. please double check the last upstream sync to ensure Brandon's build script changes are there.

**Comment by bsiegel** [comment]
Agreed with Anu about the exception feedback. I like the other comments that you have there because they tell me why that type of exception might be thrown. E.g. An `MsalServiceException` means "Exception when communicating with the STS, likely config issue".

I'd just like a similar comment that explains why an `MsalClientException` might occur. Something super brief, like "SDK was able to communicate with the STS, but was unable to authenticate successfully"

**Comment by vcolin7** [comment]
@bsiegel @anuchandy Re-applied the Gradle build fix and moved the MsalException comment to where suggested. I also added a bit more info on the other types of exception.

File: samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java [hunk 4]
```diff
@@ -40,15 +40,18 @@ public void onSuccess(IAuthenticationResult authResult) {
...
+                            // information, see: https://javadoc.io/doc/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalClientException.html
                         } else if (exception instanceof MsalServiceException) {
-                            /* Exception when communicating with the STS, likely config issue */
+                            // Exception thrown when communicating with the STS, likely because of a configuration
+                            // issue. For more information, see: https://javadoc.io/static/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalServiceException.html
                         } else if (exception instanceof MsalUiRequiredException) {
-                            /* Tokens expired or no session, retry with interactive */
+                            // Exception thrown when auth token is expired or there is no active session, must retry
+                            // with interactive login. For more information, see: https://javadoc.io/static/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalServiceException.html
```
Comment (anuchandy) [inline samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: link should be : https://javadoc.io/static/com.microsoft.identity.client/msal/0.1.1/com/microsoft/identity/client/MsalUiRequiredException.html

Comment (bsiegel) [reply samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: Good catch :)

Comment (vcolin7) [reply samples/sample-app-storage/src/main/java/com/azure/android/storage/sample/MsalClient.java]: You are right. I'll fix that real quick.

**Pull request merged**